Apps/Src/ReadSwitches.c

```diff
@@ -0,0 +1,225 @@
+//#include "Tasks.h"
+#include <os.h>
+#include "Switches.h"
+#include "Tasks.h"
+#include "CarState.h"
+
+
+/**
+ * @brief Reads Switches using Switches Driver and signals appropriate
+ * Semaphores when Switches state changes. Also updates a switches global defined in StateTypes.h, declared in Tasks.c and externed in Tasks.h
+*/
+void Task_ReadSwitches (void* p_arg) {
+    //Get Global States
+    car_state_t *CarState= (car_state_t*)p_arg;
+    switch_states_t *globalSwitchStates = &(CarState->SwitchStates);
+    switch_states_t currSwitchStates; //holds states read from driver reads
+    Switches_Init();
+
+    
+    OS_ERR err;
+    static bool arrayConThreadSpawned = false; //these booleans need to be persistent
+    static bool motorConThreadSpawned = false;
+
+    while(1){
+        // load newSwitchStates
+        currSwitchStates = (switch_states_t){
+            Switches_Read(IGN_1),
+            Switches_Read(IGN_2),
+            (velocity_switches_t){
+                Switches_Read(CRUZ_SW),
+                Switches_Read(CRUZ_EN),
+                Switches_Read(FWD_SW),
+                Switches_Read(REV_SW),
+                Switches_Read(REGEN_SW)
+            },
+            (light_switches_t){
+                Switches_Read(LEFT_SW),
+                Switches_Read(RIGHT_SW),
+                Switches_Read(HEADLIGHT_SW),
+                Switches_Read(HZD_SW)
+            }
+        };
+        
+        //Spawns arrayConnection thread and MotorConnection thread when it is appropriate
+        uint8_t ignStates = 0x00; //holds states read from Switches Driver
+        uint8_t ignStored = 0x00; //holds states stored in globals
+
+        ignStates|=((currSwitchStates.IGN_2&1)<<1); //shift IGN_2 over 1 and set it in ignStates
+        ignStates|=(currSwitchStates.IGN_1&1);
+        ignStored|=((globalSwitchStates->IGN_2&1)<<1);
+        ignStored|=(globalSwitchStates->IGN_1&1);
+
+        if (ignStored!=ignStates){ //new ignition state detected
+            switch(ignStates){
+                case 0x00:
+                    CarState->ShouldArrayBeActivated = 0;
+                    CarState->ShouldMotorBeActivated = 0;
+                    OSSemPost(
+                        &ArrayConnectionChange_Sem4,
+                        (OS_OPT) OS_OPT_POST_ALL,
+                        (OS_ERR*)&err
+                    );
+                    OSSemPost(
+                        &MotorConnectionChange_Sem4,
+                        (OS_OPT) OS_OPT_POST_ALL,
+                        (OS_ERR*)&err
+                    );
```

Comment:
Is there a reason you are posting semaphores to control the contactors instead of just calling the contactor driver directly? It would get rid of the context switch overhead and reduce the number of tasks you need. I also think it would make the code more readable and easier to debug.